File: certificate_builder.rb
# certificate_builder.rb
#!/usr/bin/env ruby
require 'json'
require 'securerandom'
require 'date'

DATA_FILE = 'templates.json'

class Template
  attr_accessor :id, :name, :description, :content, :created_at

  def initialize(name, description = '', content = '')
    @id = SecureRandom.hex(4)
    @name = name
    @description = description
    @content = content.empty? ? 'This certifies that {{name}} has completed {{course}} on {{date}}.' : content
    @created_at = Time.now.iso8601
  end

  def to_hash
    { id: @id, name: @name, description: @description, content: @content, created_at: @created_at }
  end

  def self.from_hash(h)
    t = new(h['name'], h['description'], h['content'])
    t.id = h['id']
    t.created_at = h['created_at']
    t
  end
end

class App
  attr_reader :templates

  def initialize
    @templates = []
    load
  end

  def load
    if File.exist?(DATA_FILE)
      data = JSON.parse(File.read(DATA_FILE))
      @templates = data.map { |h| Template.from_hash(h) }
    end
  end

  def save
    File.write(DATA_FILE, JSON.pretty_generate(@templates.map(&:to_hash)))
  end

  def get_template(name)
    @templates.find { |t| t.name == name }
  end

  def create(name, description = '')
    if get_template(name)
      puts "❌ Template '#{name}' already exists."
      return
    end
    t = Template.new(name, description)
    @templates << t
    save
    puts "✅ Template created: #{t.name} (ID: #{t.id})"
  end

  def list
    if @templates.empty?
      puts "No templates."
      return
    end
    puts "\n📋 Templates:"
    @templates.each_with_index do |t, i|
      desc = t.description.empty? ? 'No description' : t.description
      puts "#{i+1}. #{t.name} (#{desc})"
    end
  end

  def show(name)
    t = get_template(name)
    unless t
      puts "❌ Template '#{name}' not found."
      return
    end
    puts "\n📄 Template: #{t.name}"
    puts "Description: #{t.description.empty? ? 'None' : t.description}"
    puts "Content:"
    puts t.content
  end

  def edit(name, content)
    t = get_template(name)
    unless t
      puts "❌ Template '#{name}' not found."
      return
    end
    t.content = content
    save
    puts "✅ Template '#{name}' updated."
  end

  def generate(template_name, recipient, date = nil, course = nil, output = nil)
    t = get_template(template_name)
    unless t
      puts "❌ Template '#{template_name}' not found."
      return
    end
    date_str = date || Date.today.to_s
    course_str = course || 'the course'
    content = t.content.gsub('{{name}}', recipient)
                       .gsub('{{date}}', date_str)
                       .gsub('{{course}}', course_str)

    if output
      File.write(output, content)
      puts "✅ Certificate generated for #{recipient} -> #{output}"
    else
      puts "\n📜 Certificate for #{recipient}:\n"
      puts content
    end
  end
end

if ARGV.empty?
  puts "Usage: certificate_builder.rb <command> [options]"
  exit
end

app = App.new
cmd = ARGV.shift

case cmd
when 'template'
  sub = ARGV.shift
  case sub
  when 'create'
    name = ARGV.shift
    if name.nil?
      puts "template create <name> [--desc DESCRIPTION]"
      exit
    end
    desc = ''
    if ARGV.include?('--desc')
      idx = ARGV.index('--desc')
      desc = ARGV[idx+1] if idx
    end
    app.create(name, desc)

  when 'list'
    app.list

  when 'show'
    name = ARGV.shift
    if name.nil?
      puts "template show <name>"
      exit
    end
    app.show(name)

  when 'edit'
    if ARGV.size < 2
      puts "template edit <name> --content CONTENT"
      exit
    end
    name = ARGV.shift
    content = nil
    if ARGV.include?('--content')
      idx = ARGV.index('--content')
      content = ARGV[idx+1] if idx
    end
    if content.nil?
      puts "--content is required"
      exit
    end
    app.edit(name, content)

  else
    puts "Unknown template subcommand"
  end

when 'certificate'
  sub = ARGV.shift
  if sub != 'generate'
    puts "certificate generate <template> <recipient> [--date DATE] [--course COURSE] [--output FILE]"
    exit
  end
  template = ARGV.shift
  recipient = ARGV.shift
  if template.nil? || recipient.nil?
    puts "certificate generate requires template and recipient"
    exit
  end
  date = nil
  course = nil
  output = nil
  while ARGV.any?
    case ARGV[0]
    when '--date'
      ARGV.shift
      date = ARGV.shift
    when '--course'
      ARGV.shift
      course = ARGV.shift
    when '--output'
      ARGV.shift
      output = ARGV.shift
    else
      break
    end
  end
  app.generate(template, recipient, date, course, output)

else
  puts "Unknown command."
end

File: certificate_builder.php
# certificate_builder.php
#!/usr/bin/env php
<?php

define('DATA_FILE', 'templates.json');

class Template {
    public $id;
    public $name;
    public $description;
    public $content;
    public $created_at;

    function __construct($name, $description = '', $content = '') {
        $this->id = substr(bin2hex(random_bytes(4)), 0, 8);
        $this->name = $name;
        $this->description = $description;
        $this->content = $content ?: 'This certifies that {{name}} has completed {{course}} on {{date}}.';
        $this->created_at = date('c');
    }

    function toArray() {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'description' => $this->description,
            'content' => $this->content,
            'created_at' => $this->created_at
        ];
    }

    static function fromArray($data) {
        $t = new self($data['name'], $data['description'], $data['content']);
        $t->id = $data['id'];
        $t->created_at = $data['created_at'];
        return $t;
    }
}

class App {
    private $templates = [];

    function __construct() {
        $this->load();
    }

    function load() {
        if (file_exists(DATA_FILE)) {
            $data = json_decode(file_get_contents(DATA_FILE), true);
            $this->templates = array_map(function($d) { return Template::fromArray($d); }, $data);
        }
    }

    function save() {
        $data = array_map(function($t) { return $t->toArray(); }, $this->templates);
        file_put_contents(DATA_FILE, json_encode($data, JSON_PRETTY_PRINT));
    }

    function getTemplate($name) {
        foreach ($this->templates as $t) {
            if ($t->name == $name) return $t;
        }
        return null;
    }

    function create($name, $description = '') {
        if ($this->getTemplate($name)) {
            echo "❌ Template '$name' already exists.\n";
            return;
        }
        $t = new Template($name, $description);
        $this->templates[] = $t;
        $this->save();
        echo "✅ Template created: {$t->name} (ID: {$t->id})\n";
    }

    function list() {
        if (empty($this->templates)) {
            echo "No templates.\n";
            return;
        }
        echo "\n📋 Templates:\n";
        foreach ($this->templates as $i => $t) {
            $desc = $t->description ?: 'No description';
            echo ($i+1) . ". {$t->name} ($desc)\n";
        }
    }

    function show($name) {
        $t = $this->getTemplate($name);
        if (!$t) {
            echo "❌ Template '$name' not found.\n";
            return;
        }
        echo "\n📄 Template: {$t->name}\n";
        echo "Description: " . ($t->description ?: 'None') . "\n";
        echo "Content:\n{$t->content}\n";
    }

    function edit($name, $content) {
        $t = $this->getTemplate($name);
        if (!$t) {
            echo "❌ Template '$name' not found.\n";
            return;
        }
        $t->content = $content;
        $this->save();
        echo "✅ Template '$name' updated.\n";
    }

    function generate($templateName, $recipient, $date = null, $course = null, $output = null) {
        $t = $this->getTemplate($templateName);
        if (!$t) {
            echo "❌ Template '$templateName' not found.\n";
            return;
        }
        $dateStr = $date ?: date('Y-m-d');
        $courseStr = $course ?: 'the course';
        $content = str_replace('{{name}}', $recipient, $t->content);
        $content = str_replace('{{date}}', $dateStr, $content);
        $content = str_replace('{{course}}', $courseStr, $content);

        if ($output) {
            file_put_contents($output, $content);
            echo "✅ Certificate generated for $recipient -> $output\n";
        } else {
            echo "\n📜 Certificate for $recipient:\n\n$content\n";
        }
    }
}

if ($argc < 2) {
    die("Usage: php certificate_builder.php <command> [options]\n");
}
$app = new App();
$cmd = $argv[1];

switch ($cmd) {
    case 'template':
        if ($argc < 3) die("template: create, list, show, edit\n");
        $sub = $argv[2];
        switch ($sub) {
            case 'create':
                if ($argc < 4) die("template create <name> [--desc DESCRIPTION]\n");
                $name = $argv[3];
                $desc = '';
                for ($i=4; $i<$argc; $i++) {
                    if ($argv[$i] == '--desc' && isset($argv[$i+1])) { $desc = $argv[++$i]; }
                }
                $app->create($name, $desc);
                break;

            case 'list':
                $app->list();
                break;

            case 'show':
                if ($argc < 4) die("template show <name>\n");
                $app->show($argv[3]);
                break;

            case 'edit':
                if ($argc < 5) die("template edit <name> --content CONTENT\n");
                $name = $argv[3];
                $content = null;
                for ($i=4; $i<$argc; $i++) {
                    if ($argv[$i] == '--content' && isset($argv[$i+1])) { $content = $argv[++$i]; }
                }
                if ($content === null) die("--content is required\n");
                $app->edit($name, $content);
                break;

            default:
                echo "Unknown template subcommand\n";
        }
        break;

    case 'certificate':
        if ($argc < 4) die("certificate generate <template> <recipient> [--date DATE] [--course COURSE] [--output FILE]\n");
        $sub = $argv[2];
        if ($sub != 'generate') {
            echo "Unknown certificate subcommand\n";
            break;
        }
        $template = $argv[3];
        $recipient = $argv[4] ?? null;
        if (!$recipient) die("recipient required\n");
        $date = $course = $output = null;
        for ($i=5; $i<$argc; $i++) {
            if ($argv[$i] == '--date' && isset($argv[$i+1])) { $date = $argv[++$i]; }
            if ($argv[$i] == '--course' && isset($argv[$i+1])) { $course = $argv[++$i]; }
            if ($argv[$i] == '--output' && isset($argv[$i+1])) { $output = $argv[++$i]; }
        }
        $app->generate($template, $recipient, $date, $course, $output);
        break;

    default:
        echo "Unknown command.\n";
}
?>
